### ISSUE TITLE
Default settings for custom modal

### ISSUE THREAD

**Issue by nathasm** [shlomiassaf/angular2-modal#87]
I'm following the example for the custom modal, however, I am not sure how to set/use the default values for this scenario.

Given the example, the user will enter the answer to close the modal. However, what I would like is to use the BS defaults (with ESC being able to close the modal and/or clicking outside the modal).

Any guidance on how to enable closing the modal with ESC/clicking outside the modal?

**Comment by shlomiassaf** [comment]
The implementation is plugin specific of course but there are basic configuration details which are the `keyboard` close key and `isBlocking` ( for click ) which are set in the base `modal-context`.

Each modal has a context which is supplied when calling `open`, see [context](https://github.com/shlomiassaf/angular2-modal/blob/master/src/components/angular2-modal/models/modal-context.ts) and [open()](https://github.com/shlomiassaf/angular2-modal/blob/master/src/components/angular2-modal/providers/modal.ts#L82-L83)

For Bootstrap, the events are handled in the `modal-container` component which is the container for the custom element you provide (see [code](https://github.com/shlomiassaf/angular2-modal/blob/master/src/components/angular2-modal/plugins/bootstrap/modal-container.ts)), so it is handled for you.

then you call `modal.open(Component, Context)` the context parameter should hold the configuration for handling close keys and isBlocking

I custom modal demo in the demo app is invoked with context (see [here](https://github.com/shlomiassaf/angular2-modal/blob/master/src/components/angular2-modal/plugins/bootstrap/modal-container.ts)) which instructs what to do for keyboard/clicks.

The context code:
```ts

export class AdditionCalculateWindowData extends BSModalContext {
    constructor(public num1: number, public num2: number) {
        super();
    }
}
```
As you can see, nothing much, just inherit the base modal context for bootstrap (which inherits the base modal context in general, that has control for isBlocking and keyboard)

Changing values on a new instance of `AdditionCalculateWindowData` is what you want:
```ts
let context = new AdditionCalculateWindowData(3,2);
context.isBlocking = true; // now its blocking.

modal.open(AdditionCalculateWindow, context);
```

Note not to confuse a context class with a context **builder** class.
A builder is just a fluent api for creating context's, so you can do things like:
```ts
context.isBlocking(true).keyboard(27)......
```

This is supplied for end users via drop ins (alert, prompt, confirm)

**Comment by nathasm** [comment]
Wow, thanks for the great response. This does make things clearer for me.

In looking through your demo code, I see that the `AdditionCalculatedWindowData` does extend from `BSModalContext` so I would assume that `keyboard: [27]` and `isBlocking=false` is already set. However, neither of those will actually close the modal.

I guess my question is...is that a bug or is there another way to enable closing the modal when going the custom route?

**Comment by shlomiassaf** [comment]
No its not a bug, it's actually a feature , really :)

Every component (used in `modal.open()`) is able to control close/open actions, this is part of the interface.

So, in our case - `AdditionCalculateWindow` component is doing:
```ts
export class AdditionCalculateWindow implements ModalComponent<AdditionCalculateWindowData> {
    context: AdditionCalculateWindowData;

    public wrongAnswer: boolean;

    constructor(public dialog: DialogRef<AdditionCalculateWindowData>) {
        this.context = dialog.context;
        this.wrongAnswer = true;
    }

    onKeyUp(value) {
        this.wrongAnswer = value != 5;
        this.dialog.close();
    }


    beforeDismiss(): boolean {
        return true;
    }

    beforeClose(): boolean {
        return this.wrongAnswer;
    
```

`beforeDismiss` overrides the base implementation from `ModalComponent` (which is noop),  it **always** returns true - i.e: CANCEL the dismissal, so no matter what the settings are you can't close the modal.
However, every key press (`onKeyUp`) the correctness of the answer is checked and a close instruction is sent (to the `DialogRef`), now this call will trigger a `beforeClose` hook that will only close if `false` is returned, same as `beforeDismiss`, now if we have a `wrongAnswer` then we wont close... but else we will.

To sum up, there are 2 hooks available in a component sent to `modal.open()`:  
 * beforeDismiss
 * beforeClose

If they return true it will cancel the dismissal/close

I think too much options and advanced features creates a lot of confusion especially when there is no documentation...

**Comment by nathasm** [comment]
Great! It all makes sense now. Documentation would have certainly helped but I think the API could possibly be clearer. `shouldDismiss` or `shouldClose` might be better named? Then the return value would need to switch in order to actually close/dismiss the modal.

Thanks for a great library!
